### PR TITLE
[mlir] Initial patch to add an MPI dialect

### PR DESCRIPTION
**Pull request by AntonLydike**
This patch introduces the new MPI dialect into MLIR. The Message Passing Interface (MPI) is a widely-used standard for distributed programs to exchange data. This PR goes together with a talk later at today's LLVM Dev Meeting.

This is just a first, small patch to get going and add the necessary base files, so that we can add more operations in further patches.

Here's the documentation as generated by `ninja mlir-doc`:
# 'mpi' Dialect

This dialect models the Message Passing Interface (MPI), version 
4.0. It is meant to serve as an interfacing dialect that is targeted
by higher-level dialects. The MPI dialect itself can be lowered to 
multiple MPI implementations and hide differences in ABI. The dialect
models the functions of the MPI specification as close to 1:1 as possible
while preserving SSA value semantics where it makes sense, and uses 
`memref` types instead of bare pointers.

This dialect is under active development, and while stability is an
eventual goal, it is not guaranteed at this juncture. Given the early 
state, it is recommended to inquire further prior to using this dialect.

For an in-depth documentation of the MPI library interface, please refer 
to official documentation such as the 
[OpenMPI online documentation](https://www.open-mpi.org/doc/current/).

[TOC]

### `mpi.comm_rank` (mpi::CommRankOp)

_Get the current rank, equivalent to `MPI_Comm_rank(MPI_COMM_WORLD, &rank)`_

Syntax:

```
operation ::= `mpi.comm_rank` attr-dict `:` type(results)
```

Communicators other than `MPI_COMM_WORLD` are not supported for now.

This operation can optionally return an `!mpi.retval` value that can be used
to check for errors.

#### Results:

| Result | Description |
| :----: | ----------- |
| `retval` | MPI function call return value
| `rank` | 32-bit signless integer

### `mpi.error_class` (mpi::ErrorClassOp)

_Get the error class from an error code, equivalent to the `MPI_Error_class` function_

Syntax:

```
operation ::= `mpi.error_class` $val attr-dict `:` type($val)
```

`MPI_Error_class` maps return values from MPI calls to a set of well-known
MPI error classes.

#### Operands:

| Operand | Description |
| :-----: | ----------- |
| `val` | MPI function call return value

#### Results:

| Result | Description |
| :----: | ----------- |
| `errclass` | MPI function call return value

### `mpi.finalize` (mpi::FinalizeOp)

_Finalize the MPI library, equivalent to `MPI_Finalize()`_

Syntax:

```
operation ::= `mpi.finalize` attr-dict (`:` type($retval)^)?
```

This function cleans up the MPI state. Afterwards, no MPI methods may 
be invoked (excpet for MPI_Get_version, MPI_Initialized, and MPI_Finalized).
Notably, MPI_Init cannot be called again in the same program.

This operation can optionally return an `!mpi.retval` value that can be used
to check for errors.

#### Results:

| Result | Description |
| :----: | ----------- |
| `retval` | MPI function call return value

### `mpi.init` (mpi::InitOp)

_Initialize the MPI library, equivalent to `MPI_Init(NULL, NULL)`_

Syntax:

```
operation ::= `mpi.init` attr-dict (`:` type($retval)^)?
```

This operation must preceed most MPI calls (except for very few exceptions,
please consult with the MPI specification on these).

Passing &argc, &argv is not supported currently.

This operation can optionally return an `!mpi.retval` value that can be used
to check for errors.

#### Results:

| Result | Description |
| :----: | ----------- |
| `retval` | MPI function call return value

### `mpi.recv` (mpi::RecvOp)

_Equivalent to `MPI_Recv(ptr, size, dtype, dest, tag, MPI_COMM_WORLD, MPI_STATUS_IGNORE)`_

Syntax:

```
operation ::= `mpi.recv` `(` $ref `,` $tag `,` $rank `)` attr-dict `:` type($ref) `,` type($tag) `,` type($rank)(`->` type($retval)^)?
```

MPI_Recv performs a blocking receive of `size` elements of type `dtype` 
from rank `dest`. The `tag` value and communicator enables the library to 
determine the matching of multiple sends and receives between the same 
ranks.

Communicators other than `MPI_COMM_WORLD` are not supprted for now.
The MPI_Status is set to `MPI_STATUS_IGNORE`, as the status object 
is not yet ported to MLIR.

This operation can optionally return an `!mpi.retval` value that can be used
to check for errors.

#### Operands:

| Operand | Description |
| :-----: | ----------- |
| `ref` | memref of any type values
| `tag` | 32-bit signless integer
| `rank` | 32-bit signless integer

#### Results:

| Result | Description |
| :----: | ----------- |
| `retval` | MPI function call return value

### `mpi.retval_check` (mpi::RetvalCheckOp)

_Check an MPI return value against an error class_

Syntax:

```
operation ::= `mpi.retval_check` $val `=` $errclass attr-dict `:` type($res)
```

This operation compares MPI status codes to known error class
constants such as `MPI_SUCCESS`, or `MPI_ERR_COMM`.

#### Attributes:

<table>
<tr><th>Attribute</th><th>MLIR Type</th><th>Description</th></tr>
<tr><td><code>errclass</code></td><td>::mlir::mpi::MPI_ErrorClassEnumAttr</td><td><details><summary>MPI error class name</summary>{{% markdown %}}Enum cases:
* MPI_SUCCESS (`MPI_SUCCESS`)
* MPI_ERR_ACCESS (`MPI_ERR_ACCESS`)
* MPI_ERR_AMODE (`MPI_ERR_AMODE`)
* MPI_ERR_ARG (`MPI_ERR_ARG`)
* MPI_ERR_ASSERT (`MPI_ERR_ASSERT`)
* MPI_ERR_BAD_FILE (`MPI_ERR_BAD_FILE`)
* MPI_ERR_BASE (`MPI_ERR_BASE`)
* MPI_ERR_BUFFER (`MPI_ERR_BUFFER`)
* MPI_ERR_COMM (`MPI_ERR_COMM`)
* MPI_ERR_CONVERSION (`MPI_ERR_CONVERSION`)
* MPI_ERR_COUNT (`MPI_ERR_COUNT`)
* MPI_ERR_DIMS (`MPI_ERR_DIMS`)
* MPI_ERR_DISP (`MPI_ERR_DISP`)
* MPI_ERR_DUP_DATAREP (`MPI_ERR_DUP_DATAREP`)
* MPI_ERR_ERRHANDLER (`MPI_ERR_ERRHANDLER`)
* MPI_ERR_FILE (`MPI_ERR_FILE`)
* MPI_ERR_FILE_EXISTS (`MPI_ERR_FILE_EXISTS`)
* MPI_ERR_FILE_IN_USE (`MPI_ERR_FILE_IN_USE`)
* MPI_ERR_GROUP (`MPI_ERR_GROUP`)
* MPI_ERR_INFO (`MPI_ERR_INFO`)
* MPI_ERR_INFO_KEY (`MPI_ERR_INFO_KEY`)
* MPI_ERR_INFO_NOKEY (`MPI_ERR_INFO_NOKEY`)
* MPI_ERR_INFO_VALUE (`MPI_ERR_INFO_VALUE`)
* MPI_ERR_IN_STATUS (`MPI_ERR_IN_STATUS`)
* MPI_ERR_INTERN (`MPI_ERR_INTERN`)
* MPI_ERR_IO (`MPI_ERR_IO`)
* MPI_ERR_KEYVAL (`MPI_ERR_KEYVAL`)
* MPI_ERR_LOCKTYPE (`MPI_ERR_LOCKTYPE`)
* MPI_ERR_NAME (`MPI_ERR_NAME`)
* MPI_ERR_NO_MEM (`MPI_ERR_NO_MEM`)
* MPI_ERR_NO_SPACE (`MPI_ERR_NO_SPACE`)
* MPI_ERR_NO_SUCH_FILE (`MPI_ERR_NO_SUCH_FILE`)
* MPI_ERR_NOT_SAME (`MPI_ERR_NOT_SAME`)
* MPI_ERR_OP (`MPI_ERR_OP`)
* MPI_ERR_OTHER (`MPI_ERR_OTHER`)
* MPI_ERR_PENDING (`MPI_ERR_PENDING`)
* MPI_ERR_PORT (`MPI_ERR_PORT`)
* MPI_ERR_PROC_ABORTED (`MPI_ERR_PROC_ABORTED`)
* MPI_ERR_QUOTA (`MPI_ERR_QUOTA`)
* MPI_ERR_RANK (`MPI_ERR_RANK`)
* MPI_ERR_READ_ONLY (`MPI_ERR_READ_ONLY`)
* MPI_ERR_REQUEST (`MPI_ERR_REQUEST`)
* MPI_ERR_RMA_ATTACH (`MPI_ERR_RMA_ATTACH`)
* MPI_ERR_RMA_CONFLICT (`MPI_ERR_RMA_CONFLICT`)
* MPI_ERR_RMA_FLAVOR (`MPI_ERR_RMA_FLAVOR`)
* MPI_ERR_RMA_RANGE (`MPI_ERR_RMA_RANGE`)
* MPI_ERR_RMA_SHARED (`MPI_ERR_RMA_SHARED`)
* MPI_ERR_RMA_SYNC (`MPI_ERR_RMA_SYNC`)
* MPI_ERR_ROOT (`MPI_ERR_ROOT`)
* MPI_ERR_SERVICE (`MPI_ERR_SERVICE`)
* MPI_ERR_SESSION (`MPI_ERR_SESSION`)
* MPI_ERR_SIZE (`MPI_ERR_SIZE`)
* MPI_ERR_SPAWN (`MPI_ERR_SPAWN`)
* MPI_ERR_TAG (`MPI_ERR_TAG`)
* MPI_ERR_TOPOLOGY (`MPI_ERR_TOPOLOGY`)
* MPI_ERR_TRUNCATE (`MPI_ERR_TRUNCATE`)
* MPI_ERR_TYPE (`MPI_ERR_TYPE`)
* MPI_ERR_UNKNOWN (`MPI_ERR_UNKNOWN`)
* MPI_ERR_UNSUPPORTED_DATAREP (`MPI_ERR_UNSUPPORTED_DATAREP`)
* MPI_ERR_UNSUPPORTED_OPERATION (`MPI_ERR_UNSUPPORTED_OPERATION`)
* MPI_ERR_VALUE_TOO_LARGE (`MPI_ERR_VALUE_TOO_LARGE`)
* MPI_ERR_WIN (`MPI_ERR_WIN`)
* MPI_ERR_LASTCODE (`MPI_ERR_LASTCODE`){{% /markdown %}}</details></td></tr>
</table>

#### Operands:

| Operand | Description |
| :-----: | ----------- |
| `val` | MPI function call return value

#### Results:

| Result | Description |
| :----: | ----------- |
| `res` | 1-bit signless integer

### `mpi.send` (mpi::SendOp)

_Equivalent to `MPI_Send(ptr, size, dtype, dest, tag, MPI_COMM_WORLD)`_

Syntax:

```
operation ::= `mpi.send` `(` $ref `,` $tag `,` $rank `)` attr-dict `:` type($ref) `,` type($tag) `,` type($rank)(`->` type($retval)^)?
```

MPI_Send performs a blocking send of `size` elements of type `dtype` to rank
`dest`. The `tag` value and communicator enables the library to determine 
the matching of multiple sends and receives between the same ranks.

Communicators other than `MPI_COMM_WORLD` are not supprted for now.

This operation can optionally return an `!mpi.retval` value that can be used
to check for errors.

#### Operands:

| Operand | Description |
| :-----: | ----------- |
| `ref` | memref of any type values
| `tag` | 32-bit signless integer
| `rank` | 32-bit signless integer

#### Results:

| Result | Description |
| :----: | ----------- |
| `retval` | MPI function call return value

### MPI_ErrorClassEnumAttr

MPI error class name

Syntax:

```
#mpi.errclass<
  ::mlir::mpi::MPI_ErrorClassEnum   # value
>
```

Enum cases:
* MPI_SUCCESS (`MPI_SUCCESS`)
* MPI_ERR_ACCESS (`MPI_ERR_ACCESS`)
* MPI_ERR_AMODE (`MPI_ERR_AMODE`)
* ... *all other MPI error codes*

#### Parameters:

| Parameter | C++ type | Description |
| :-------: | :-------: | ----------- |
| value | `::mlir::mpi::MPI_ErrorClassEnum` | an enum of type MPI_ErrorClassEnum |

### RetvalType

MPI function call return value

Syntax: `!mpi.retval`

This type represents a return value from an MPI function vall.
This value can be MPI_SUCCESS, MPI_ERR_IN_STATUS, or any error code.

This return value can be compared agains the known MPI error classes
represented by `#mpi.errclass` using the `mpi.retval_check` operation.

### CODE REVIEW

**Comment by luisacicolini** [comment]
This looks interesting! I'm curious to see how it works in detail.

**Comment by kiranchandramohan** [comment]
The Flang community is also interested in this work. Would you be interested in presenting at our regular calls? We have calls every week (The next occurrences are on Monday 16th and Wednesday 25th). Thanks in advance.

**Comment by xblang-project** [comment]
This looks great, thank you for working on this!
I think an RFC would be appreciated, especially to discuss future development. In this PR adding more test cases, including invalid syntax ones would be great.
Great work again!

**Comment by fabianmcg** [comment]
This looks great, thank you for working on this!
I think an RFC would be appreciated, especially to discuss future development. In this PR adding more test cases, including invalid syntax ones would be great.
Great work again!

**Comment by clementval** [comment]
> I think an RFC would be appreciated, especially to discuss future development. 
+1

**Comment by AntonLydike** [comment]
@fabianmcg @clementval, the next ODM no Nov. 2nd will have a presentation and discussion round on the MPI dialect!

**Comment by clementval** [comment]
A RFC on discord would still be appreciated to discuss the advantage of having a MPI dialect.

**Comment by AntonLydike** [comment]
RFC has been posted here: https://discourse.llvm.org/t/rfc-mpi-dialect/74705

**Comment by fschlimb** [comment]
+1
Excellent!
I suggest adding returning error codes right away.

**Comment by AntonLydike** [comment]
Thanks for the feedback @fschlimb, I'll see that I get an updated PR ready for review ASAP!

**Comment by rengolin** [comment]
The failure in `Lower/OpenMP/parallel-private-clause-fixes.f90` seems unrelated and looks just like bad test (`CHECK-DAG` instead of `CHECK`). I think I've seen this error before in other unrelated PRs: @llvm/pr-subscribers-flang-fir-hlfir

**Comment by kiranchandramohan** [comment]
> The failure in `Lower/OpenMP/parallel-private-clause-fixes.f90` seems unrelated and looks just like bad test (`CHECK-DAG` instead of `CHECK`). I think I've seen this error before in other unrelated PRs: @llvm/pr-subscribers-flang-fir-hlfir

Will push a fix soon. Thanks for bringing it to our attention.

File: mlir/include/mlir/Dialect/MPI/IR/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,25 @@
+add_mlir_dialect(MPI mpi)
+add_mlir_doc(MPI MPI Dialects/ -gen-dialect-doc)
+
+# Add MPI operations
+set(LLVM_TARGET_DEFINITIONS MPIOps.td)
+mlir_tablegen(MPIOps.h.inc -gen-op-decls)
+mlir_tablegen(MPIOps.cpp.inc -gen-op-defs)
+add_public_tablegen_target(MLIRMPIOpsIncGen)
+add_dependencies(mlir-generic-headers MLIRMPIOpsIncGen)
```
Comment (Groverkss) [inline mlir/include/mlir/Dialect/MPI/IR/CMakeLists.txt:9]: mlir-generic-headers are dialect independent. I don't think we should be adding a dependency of MPI dialect on them.

File: mlir/include/mlir/Dialect/MPI/IR/MPI.td [hunk 2]
```diff
@@ -0,0 +1,217 @@
+//===- MPIBase.td - Base defs for mpi dialect ---------------*- tablegen -*-==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef MLIR_DIALECT_MPI_IR_MPI
```
Comment (Groverkss) [inline mlir/include/mlir/Dialect/MPI/IR/MPI.td:9]: nit: MLIR_DIALECT_MPI_IR_MPI_TD

File: mlir/include/mlir/Dialect/MPI/IR/MPIOps.td [hunk 3]
```diff
@@ -0,0 +1,196 @@
+//===- MPI.td - Message Passing Interface Ops --------------*- tablegen -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef MPI_MLIR_IR_MPIOPS
```
Comment (Groverkss) [inline mlir/include/mlir/Dialect/MPI/IR/MPIOps.td:9]: nit: _TD

File: mlir/include/mlir/Dialect/MPI/IR/MPIOps.td [hunk 4]
```diff
@@ -0,0 +1,196 @@
+//===- MPI.td - Message Passing Interface Ops --------------*- tablegen -*-===//
```
Comment (Groverkss) [inline mlir/include/mlir/Dialect/MPI/IR/MPIOps.td:1]: Wrong file name

File: mlir/include/mlir/Dialect/MPI/IR/MPITypes.td [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file declares the Message Passing Interface dialect types.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef MLIR_DIALECT_MPI_IR_MPITYPES
```
Comment (Groverkss) [inline mlir/include/mlir/Dialect/MPI/IR/MPITypes.td:13]: nit: _TD

File: mlir/include/mlir/Dialect/MPI/IR/MPITypes.td [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+#define MLIR_DIALECT_MPI_IR_MPITYPES
+
+include "mlir/IR/AttrTypeBase.td"
+include "mlir/Dialect/MPI/IR/MPI.td"
+
+//===----------------------------------------------------------------------===//
+// NPI Types
```
Comment (Groverkss) [inline mlir/include/mlir/Dialect/MPI/IR/MPITypes.td:20]: nit: MPI

File: mlir/include/mlir/Dialect/MPI/IR/MPITypes.td [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+//===----------------------------------------------------------------------===//
+
+class MPI_Type<string name, string typeMnemonic, list<Trait> traits = []>
+    : TypeDef<MPI_Dialect, name, traits> {
+  let mnemonic = typeMnemonic;
+}
+
+//===----------------------------------------------------------------------===//
+// pdl::AttributeType
```
Comment (Groverkss) [inline mlir/include/mlir/Dialect/MPI/IR/MPITypes.td:29]: Wrong name

File: mlir/lib/Dialect/MPI/IR/MPIOps.cpp [hunk 8]
```diff
@@ -0,0 +1,21 @@
+//===- MPIOps.cpp - MPI dialect ops implementation ------------===//
```
Comment (Groverkss) [inline mlir/lib/Dialect/MPI/IR/MPIOps.cpp:1]: nit: Extend -----===// to reach end of char limit

File: mlir/test/Dialect/MPI/ops.mlir [hunk 9]
```diff
@@ -0,0 +1,36 @@
+// RUN: mlir-opt %s | mlir-opt | FileCheck %s
```
Comment (Groverkss) [inline mlir/test/Dialect/MPI/ops.mlir:1]: Mention why you are using mlir-opt twice.

**Comment by AntonLydike** [comment]
Thank you @Groverkss for catching these! I've updated the PR to reflect your feedback!

File: mlir/test/Dialect/MPI/ops.mlir (hunk 9 above)
Comment (AntonLydike) [reply mlir/test/Dialect/MPI/ops.mlir:2]: I added a comment above, I'm not 100% on if this is the best approach

Comment (Dinistro) [reply mlir/test/Dialect/MPI/ops.mlir:2]: Drive by comment: The generic form is not necessary, as it by definition uses the generic parser.

File: mlir/include/mlir/Dialect/MPI/IR/CMakeLists.txt (hunk 1 above)
Comment (Groverkss) [reply mlir/include/mlir/Dialect/MPI/IR/CMakeLists.txt]: I don't think this was resolved

Comment (AntonLydike) [reply mlir/include/mlir/Dialect/MPI/IR/CMakeLists.txt]: Good catch! Cleaned the last one up too!

File: mlir/test/Dialect/MPI/ops.mlir (hunk 9 above)
Comment (AntonLydike) [reply mlir/test/Dialect/MPI/ops.mlir:1]: I changed it back to the default round-trip test, thanks for the input @Dinistro!

**Review by Dinistro — COMMENTED** [review]
Dropped a bunch of nit comments, the general design makes a lot of sense to me.

One concern I'm having is the restriction to only work with memref types. While I understand that this is easier to work with, some users might not be able to work with memrefs and thus cannot use this dialect. AFAIK, flang doesn't use memrefs, and neither does code that will come from some lower level input (or the LLVM import).
Note that this can also be resolved once this has been landed, though.

I think this is looking good to me, but I don't have the authority to accept this, I think. @kiranchandramohan @rengolin any idea who has the authority?

File: mlir/include/mlir/Dialect/MPI/IR/MPITypes.td [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+//===----------------------------------------------------------------------===//
+// mpi::RetvalType
+//===----------------------------------------------------------------------===//
+
+def MPI_Retval : MPI_Type<"Retval", "retval"> {
+  let summary = "MPI function call return value";
+  let description = [{This type represents a value returned from an MPI
+                          call.This value can be MPI_SUCCESS,
+                      MPI_ERR_IN_STATUS, or any error code.}];
```
[the remark this answers is not in the document]
Comment (AntonLydike) [reply mlir/include/mlir/Dialect/MPI/IR/MPITypes.td]: This seems to be a weird artifact of `clang-format`. I asked for help on [discord](https://discord.com/channels/636084430946959380/642426447167881246/1193190420973375518).

For now I manually ignored `clang-format`s hints for this file.

File: mlir/test/Dialect/MPI/invalid.mlir
[the remark this answers is not in the document]
Comment (AntonLydike) [reply mlir/test/Dialect/MPI/invalid.mlir:1]: Thanks for the tip! After some consideration and feedback we decided to drop the file entirely, as we only have the verifies generated by tablegen in the dialect.

**Comment by Groverkss** [comment]
This LGTM. Let's wait another week since I expect folks are still returning from vacation.

There seems to be enough interest in the MPI Dialect, and I don't see any outstanding concerns, so I'm happy to accept it after a week.

Tagging people who initially showed interest to see if there are any outstanding concerns @rengolin @nicolasvasilache @kiranchandramohan @clementval @fschlimb @fabianmcg @sogartar @tschuett @yaochengji

**Comment by kiranchandramohan** [comment]
Could you summarise the answers to the questions in the guidelines on contributing a new dialect in the RFC? Apologies, if this has been done already, then a link will be useful.

https://mlir.llvm.org/getting_started/DeveloperGuide/#guidelines-on-contributing-a-new-dialect-or-important-components


> One concern I'm having is the restriction to only work with memref types. While I understand that this is easier to work with, some users might not be able to work with memrefs and thus cannot use this dialect. AFAIK, flang doesn't use memrefs, and neither does code that will come from some lower level input (or the LLVM import).

Yes, a memref only approach will make it not immediately useful in Flang. But that need not be a blocker. Fortran has co-arrays which can potentially lower to MPI.


> The failure in `Lower/OpenMP/parallel-private-clause-fixes.f90` seems unrelated and looks just like bad test (`CHECK-DAG` instead of `CHECK`).

Is this still failing? I thought I had fixed in https://github.com/llvm/llvm-project/commit/a4deb14e353cd8c543f78979ce986f98023d36dd.

**Comment by AntonLydike** [comment]
As @kiranchandramohan said:
> Could you summarise the answers to the questions in the guidelines on contributing a new dialect in the RFC? Apologies, if this has been done already, then a link will be useful.
> 
> https://mlir.llvm.org/getting_started/DeveloperGuide/#guidelines-on-contributing-a-new-dialect-or-important-components

Here are our answers as requested:

> What is the overall goal of the dialect?

An MPI dialect will provide first-class message-passing primitives that can be used both as a target by higher-level code generation and as a fixed point for other lower-level efforts such as MPI-level optimizations and rewrites.

> What is the first implementation milestone?

As a first implementation milestone, we will introduce ops for blocking communication and a lowering to llvm.

> How does it fit into the MLIR dialect ecosystem?

It fits in as a target-able dialect that abstracts away the messy MPI ABI and helps bridge the gap between pointers and MLIR's memref.

> Connection: how does it connect to the existing dialects in a compilation pipeline(s)?

The MPI dialect will work as a target for higher-level message-passing abstractions such as the sharding dialect. The lower end of a sharded linalg flow would encompass (with more intermediate steps) (linalg + sharding) -> (scf + memref + MPI) -> (llvm). We have demonstrated this approach by building a stencil compiler on a slightly different higher level message passing abstraction that goes (stencil) -> (stencil + message passing) -> (scf + memref + MPI) -> (llvm). 

> Consolidation: is there already a dialect with a similar goal or matching abstractions; if so, can it be improved instead of adding a new one?

We are unaware of any other dialect in MLIR filling a similar role to this one.

> Reuse: how does it generalize to similar but slightly different use cases?

While we focus on a target dialect for HPC code generation, other groups are interested in exploring a different space, such as MPI-level optimizations or translating point-to-point MPI calls to other forms of communication. We are confident that our abstractions can generalize to these use cases without a significant effort. We have collaborated with Jeff Hammond from the MPI ABI standardization body to ensure that our abstractions are a sensible expression of the MPI Interface.

> What is the community of users that it is serving?

The proposed MPI dialect makes MLIR accessible to the HPC community and provides MLIR's AI compiler community access to HPC-style distributed computing. MPI is a broadly established communication standard used in HPC and almost uniformly used as its communication abstraction.

> Who are the future contributors/maintainers beyond those who propose the dialect?

HPC DSL developers such as Frank Schlimbach at Intel are actively looking into MLIR, and Renato Golin at Intel agreed to co-maintain the dialect with us. Unlocking access to the MLIR ecosystem through the MPI dialect will ensure they can contribute to this dialect as part of their work, ensuring that the dialect will receive the needed support.

**Comment by kiranchandramohan** [comment]
> As @kiranchandramohan said:
> 
> > Could you summarise the answers to the questions in the guidelines on contributing a new dialect in the RFC? Apologies, if this has been done already, then a link will be useful.
> > https://mlir.llvm.org/getting_started/DeveloperGuide/#guidelines-on-contributing-a-new-dialect-or-important-components
> 
> Here are our answers as requested:
> 

Thanks for the detailed answer. Could you post a copy of this in the RFC? Also, tag the people whom you have mentioned will help maintain.

File: mlir/include/mlir/Dialect/MPI/IR/MPITypes.td [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+//===----------------------------------------------------------------------===//
+// mpi::RetvalType
+//===----------------------------------------------------------------------===//
+
+def MPI_Retval : MPI_Type<"Retval", "retval"> {
+  let summary = "MPI function call return value";
+  let description = [{
+    This type represents a return value from an MPI function vall.
```
Comment (sjw36) [inline mlir/include/mlir/Dialect/MPI/IR/MPITypes.td:35]: typo: `vall` -> `call`

**Review by sjw36 — APPROVED** [review]
Thanks for this effort! Looks like a good start.
Looking forward to the non-blocking PR! Thanks.

File: mlir/include/mlir/Dialect/MPI/IR/MPI.td [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+    This dialect is under active development, and while stability is an
+    eventual goal, it is not guaranteed at this juncture. Given the early 
+    state, it is recommended to inquire further prior to using this dialect.
+
+    For an in-depth documentation of the MPI library interface, please refer 
+    to official documentation such as the 
+    [OpenMPI online documentation](https://www.open-mpi.org/doc/current/).
+  }];
+
+  let usePropertiesForAttributes = 1;
```
Comment (joker-eph) [inline mlir/include/mlir/Dialect/MPI/IR/MPI.td:37]: Nit: this is the default.

**Review by joker-eph — APPROVED** [review]
LG, please rebase and re-run the CI though.

**Comment by AntonLydike** [comment]
Thank you @sjw36 and @joker-eph for your reviews, I addressed your feedback in the latest update to this PR

**Comment by Groverkss** [comment]
@AntonLydike I reverted this for now https://github.com/llvm/llvm-project/pull/81884 . There seem to be buildbot failures.

File: mlir/lib/Dialect/MPI/IR/MPI.cpp [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+#define GET_TYPEDEF_CLASSES
+#include "mlir/Dialect/MPI/IR/MPITypesGen.cpp.inc"
+
+#include "mlir/Dialect/MPI/IR/MPIEnums.cpp.inc"
+
+#define GET_ATTRDEF_CLASSES
+#include "mlir/Dialect/MPI/IR/MPIAttrDefs.cpp.inc"
+
+#define GET_OP_CLASSES
+#include "mlir/Dialect/MPI/IR/MPIOps.cpp.inc"
```
Comment (joker-eph) [inline mlir/lib/Dialect/MPI/IR/MPI.cpp:56]: Seems like these two lines should be removed

**Pull request merged**